### PR TITLE
✨ Enable configuring Swagger UI parameters

### PR DESCRIPTION
**Pull request by jmriebold**
At present there are a number of [Swagger UI parameters](https://swagger.io/docs/open-source-tools/swagger-ui/usage/configuration/) that are not configurable when using FastAPI. Under normal circumstances most of these wouldn't need to be used at all, however there are some cases where you might need to override some of these. For example, Swagger UI's syntax highlighting is [known to be slow](https://github.com/swagger-api/swagger-ui/issues/3832), and being able to easily disable this would be very useful in certain situations.

### CODE REVIEW

File: docs/en/docs/advanced/extending-openapi.md [hunk 1]
```diff
@@ -248,3 +248,16 @@ Now, to be able to test that everything works, create a *path operation*:
 Now, you should be able to disconnect your WiFi, go to your docs at <a href="http://127.0.0.1:8000/docs" class="external-link" target="_blank">http://127.0.0.1:8000/docs</a>, and reload the page.
 
 And even without Internet, you would be able to see the docs for your API and interact with it.
+
+## Configuring Swagger UI
+
+Extra [Swagger UI parameters](https://swagger.io/docs/open-source-tools/swagger-ui/usage/configuration) can be set using the `swagger_ui_parameters` argument either in the `FastAPI()` constructor directly in the `
+get_swagger_ui_html()` method. This can be used to disable Swagger UI's syntax highlighting, change the theme, etc.
+
+`swagger_ui_parameters` must a dict mapping strings to strings, as described in the Swagger UI documentation. For example, to disable syntax highlighting:
```
Comment (pablogamboa) [inline docs/en/docs/advanced/extending-openapi.md:257]: typo: `must be a dict`

Comment (jmriebold) [reply docs/en/docs/advanced/extending-openapi.md]: Good catch, thanks!

**Comment by jmriebold** [comment]
@pablogamboa, any other changes needed to wrap this one up?

**Review by pablogamboa — APPROVED** [review]
lgtm great work @jmriebold

**Comment by jmriebold** [comment]
Thanks, appreciate the review @pablogamboa!

**Comment by long2ice** [comment]
Look forward for this merge!

File: fastapi/openapi/docs.py [hunk 2]
```diff
@@ -14,6 +14,7 @@ def get_swagger_ui_html(
     swagger_favicon_url: str = "https://fastapi.tiangolo.com/img/favicon.png",
     oauth2_redirect_url: Optional[str] = None,
     init_oauth: Optional[Dict[str, Any]] = None,
+    swagger_ui_parameters: Optional[Dict[str, str]] = None,
```
Comment (akihiro-inui) [inline fastapi/openapi/docs.py:17]: I think the second parameter of this line `Dict[str, str]` can be also an integer so I suggest to change it into  `Dict[str, any] `

**Review by akihiro-inui — CHANGES_REQUESTED** [review]
swagger_ui_parameters argument type

File: fastapi/openapi/docs.py (hunk 2 above)
Comment (jmriebold) [reply fastapi/openapi/docs.py]: Good point, changed.

Comment (akihiro-inui) [reply fastapi/openapi/docs.py]: Nice!

**Review by akihiro-inui — APPROVED** [review]

**Review by ycd — APPROVED** [review]

**Comment by jmriebold** [comment]
@pablogamboa, @akihiro-inui, @ycd, this PR's been approved for several weeks now, could we get this merged? Do you need anything else from me?

**Comment by ycd** [comment]
Nope, the change looks good, but only @tiangolo has access to master branch. We need to wait his input.

**Comment by jmriebold** [comment]
Understood, thanks @ycd!

**Comment by c-shekhar** [comment]
Hey, @tiangolo, any plans to get this PR merged soon?
Also @jmriebold great addition! :100: 
We are having a use-case where we are exposing alpha release of our APIs to customers and `tryItOutEnabled` display configuration of [swagegr-ui](https://github.com/swagger-api/swagger-ui/blob/master/docs/usage/configuration.md#display) will be very helpful and handy there for a particular group of end users.

**Comment by stvnksslr** [comment]
Bumping this as it would be a huge help to my teams everyday workflow specifically the syntax highlighting being slow for even fairly small datasets is a killer.

**Comment by jziolkowski** [comment]
another :+1: for this bump from our team as we needed to add some ugly hacks to achieve the same result, while we could have this builtin so nicely. Great job @jmriebold !

File: fastapi/openapi/docs.py [hunk 3]
```diff
@@ -34,6 +35,10 @@ def get_swagger_ui_html(
         url: '{openapi_url}',
     """
 
+    if swagger_ui_parameters:
+        for key, value in swagger_ui_parameters.items():
+            html += f"{key}: {value},"
```
Comment (elrik75) [inline fastapi/openapi/docs.py:40]: To follow the current JS rendering, you should add an `\n` at the end.
I don't think it's a good idea to pass JS code in argument `{value}`. It's easy to break all the javascript rendering in the browser. Why not using JSON to escape?
```
>>> json.dumps(False)
'false'
>>> json.dumps("string")
'"string"'
>>> json.dumps(1)
'1'
```

Comment (jmriebold) [reply fastapi/openapi/docs.py]: Great point, I've made the changes and rebased onto master.

File: fastapi/openapi/docs.py [hunk 4]
```diff
@@ -34,6 +35,10 @@ def get_swagger_ui_html(
         url: '{openapi_url}',
     """
 
+    if swagger_ui_parameters:
+        for key, value in swagger_ui_parameters.items():
+            html += f"{key}: {json.dumps(value)},\n"
```
Comment (elrik75) [inline fastapi/openapi/docs.py:40]: Maybe, we could catch the `TypeError` we will have ifthe value is not serializable in JSON (datetime, object...) to facilitate the dev experience. Here a proposition:

```python
if swagger_ui_parameters:
    try:
        ...
    except TypeError:
        print("swagger_ui values are serialized to Javascript code, please use only basic types like strings, numbers and booleans"")
        raise

Comment (jmriebold) [reply fastapi/openapi/docs.py]: Good idea, change made.

**Comment by aperullo** [comment]
What's holding this MR up?

**Comment by jmriebold** [comment]
@aperullo, as I understand it @tiangolo is the only one with write-access, so it's blocked on him reviewing and/merging it.

**Comment by elrik75** [comment]
And @tiangolo seems to be focused on other projects now. SQLModel and Typed. FastAPI needs more maintainers.

**Comment by Armster15** [comment]
@tiangolo is there a specific reason this hasn't gotten merged yet? This PR has been approved by the reviewers, and this is something many people want (personally me as well 😄). I find it a little odd that there is no official response from you yet.

I hope this didn't come off as condescending by the way, thanks so much for your work on FastAPI :)

**Comment by ghandic** [comment]
I've put in a PR for a plug-in system too, I think it would be better to have an approach that is more general and flexible

**Comment by long2ice** [comment]
This project probably don't like new feature, check the recent activity

**Comment by alexjolig** [comment]
I'm having a big issue with large response JSONs and because I'm not able to deactivate syntax highlights from Swagger UI, my project flow is freezing. Merging this would really keeps me going. @tiangolo I hope you're somewhere to be found!

**Comment by tiangolo** [comment]
Thanks for your work @jmriebold! :cake: 

And thanks for the discussion everyone.

I fixed an error in the implementation that would break Swagger UI if parameters with dots like `syntaxHighlight.theme` were used.

I also refactored the other default configurations included by FastAPI to allow overriding them, and documented it.

I refactored and updated the docs, to include source code examples that are now tested, explain defaults and overrides, and added some screenshots.

This will be available in FastAPI version `0.72.0`, I'll release it in a couple of hours. 🔖

**Comment by jmriebold** [comment]
Thanks @tiangolo!

**Comment by ghandic** [comment]
This will now require rework to the plug-in system, is there still interest in having that? Don't want to reinvest the time if there isn't the appetite https://github.com/tiangolo/fastapi/pull/4060

**Pull request merged**